File: c/sublist/sublist.c
// exercism sublist -- are either of two lists sublists
// of the other?
//
// list a is a sublist of list b if by dropping 0 or more
// elements from the front and/or back, you get a list
// that's completely equal to list a.
//
// t.brumley, june 2022.

#include "sublist.h"
#include <stdbool.h>

// forward definitions
static
bool areEqual(
		int *left,
		int *right,
		size_t nLeft,
		size_t nRight);

static
bool containedIn(
		int *left,
		int *right,
		size_t nLeft,
		size_t nRight);

// is the left list equal to the right?
static
bool areEqual(
		int *left,
		int *right,
		size_t nLeft,
		size_t nRight) {
	if (nLeft != nRight)
		return false;
	int *x, *y;
	x = left;
	y = right;
	for(size_t i = 0; i < nLeft; i++)
		if (*x++ != *y++)
			return false;
	return true;
}

// is the left list contained in the right?
static
bool containedIn(
		int *left,
		int *right,
		size_t nLeft,
		size_t nRight) {

	if (areEqual(left, right, nLeft, nRight))
		return true;

	if (nLeft == 0 && nRight != 0)
		return true;

	if (nLeft > nRight)
		return false;

	// match left against first nLeft of right, checking for equal
	while (nRight >= nLeft) {
		if (areEqual(left, right, nLeft, nLeft))
			return true;
		right++;
		nRight--;
	}

	return false;
}

// api for the comparison, lists are integer arrays with length
// explicitly passed.
comparison_result_t check_lists(
		int *compareList,
		int *baseList,
		size_t nCompareList,
		size_t nBaseList) {

	// if the lists are the same length, then this is a straight equality check
	if (nCompareList == nBaseList) {
		if (areEqual(compareList, baseList, nCompareList, nBaseList))
			return EQUAL;
		else
			return UNEQUAL;
	}

	// if one list is empty and the other is not, the empty is a sublist of the other
	if (nCompareList == 0)
		return SUBLIST;
	if (nBaseList == 0)
		return SUPERLIST;

	// both lists have elements, see if one is contained in the other
	if (containedIn(compareList, baseList, nCompareList, nBaseList))
		return SUBLIST;
	if (containedIn(baseList, compareList, nBaseList, nCompareList))
		return SUPERLIST;

	// no hope
	return UNEQUAL;
}

File: go/sublist/sublist.go
package sublist

// Type relation is a synonym for string. No validation
// of the allowed values is done. The values are:
// equal, sublist, superlist, unequal
// type Relation string

// Sublist examines the relationship between two lists
// and returns that Relation.
func Sublist(l1, l2 []int) Relation {
	// quick guards
	if len(l1) == 0 && len(l2) == 0 {
		return RelationEqual
	}
	if len(l1) == 0 && len(l2) != 0 {
		return RelationSublist
	}
	if len(l1) != 0 && len(l2) == 0 {
		return RelationSuperlist
	}

	// simple case, lists of the same length
	// either are or aren't equal
	if slicesEqual(l1, l2) {
		return RelationEqual
	}

	// sublist & superlist checks
	if len(l1) < len(l2) {
		for i := 0; i <= len(l2)-len(l1); i++ {
			if slicesEqual(l1, l2[i:len(l1)+i]) {
				return RelationSublist
			}
		}
	} else {
		for i := 0; i <= len(l1)-len(l2); i++ {
			if slicesEqual(l2, l1[i:len(l2)+i]) {
				return RelationSuperlist
			}
		}
	}
	return RelationUnequal

}

// Function slicesEqual compares two slices.
func slicesEqual(l1 []int, l2 []int) bool {
	if len(l1) != len(l2) {
		return false
	}
	for i := 0; i < len(l1); i++ {
		if l1[i] != l2[i] {
			return false
		}
	}
	return true
}
